Convert host <-> guest communication to use MessageChannel

robertknight opened the pull request:
This commit implements the initial transition of communication between
guest/host frames and the sidebar to use `MessageChannel` as the
underlying transport instead of `window.postMessage`, building on
previous changes to the `Bridge` and `RPC` classes.

This change aligns with an ongoing to plan to use `MessageChannel`
instead of `window.postMessage`, and also resolves an issue where guest
frames may fail to connect to the sidebar if the annotator code loads in
the guest before the sidebar application has finished loading.

This initial version has several limitations compared to what is planned
for the final version of the new inter-frame communication system:

 - It only supports guest frames which are the same frame as the host or
   a direct, same-origin child.
 - Sidebar <-> host communication relies on the host frame also being a
   guest frame. In order for the host frame to receive messages from the
   sidebar, it must run the logic to establish sidebar <-> guest communication.
   In other words, it is not possible to have a host frame which cannot
   also be annotated.
 - The only supported roles for frames are sidebar and host/guest. There
   is no separate role for notebook frames for use in notebook <->
   sidebar communication.

Making this change involved replacing the protocol used by guest frames
and the sidebar to discover each other. The new one is a minimal protocol
which is intended to be temporary and replaced with a different one in future
that addresses the above limitations (see `shared/communicator.js` in [this draft](https://github.com/hypothesis/client/pull/3534/files)).

1. When the sidebar application starts up it notifies the parent frame that
   it is ready to connect to guests via a `hypothesisSidebarReady`
   message.
2. Guest frames wait for this notification to be received before
   connecting to the sidebar. When they connect, they create a
   `MessageChannel` and send one port to the sidebar via a
   `hypothesisGuestReady` message and use the other port locally.
3. When the sidebar receives a `hypothesisGuestReady` message it creates
   a channel to communicate with the guest.

esanzgar commented on src/sidebar/services/frame-sync.js:
```diff
@@ -235,17 +236,30 @@ export class FrameSyncService {
+      if (e.data?.type !== 'hypothesisGuestReady') {
+        return;
+      }
+      const port = /** @type {unknown} */ (e.data.port);
```
There is no need to transferred the port on the `data`. You can reach directly from `e.ports[0]`.

https://github.com/mdn/dom-examples/blob/92687307cb13a922e4a3ef1cc16d36fe64fcec28/channel-messaging-basic/page2.html#L24

robertknight replied:
Ah, thanks for the tip.

esanzgar commented on src/sidebar/services/frame-sync.js:
```diff
@@ -235,17 +236,30 @@ export class FrameSyncService {
+    // The message will include a `MessagePort` to use for communication with
+    // the guest. Communication with the host currently relies on the host
+    // frame also always being a guest frame.
+    this._window.addEventListener('message', e => {
```
This listener is not removed. It may not be necessary, but it is always a good idea. Is there a way to cleanly destroy services, if so `ListenerCollection` utility could be handy.

esanzgar replied:
If the listener is not destroyed, this class will never be able to be garbage collected, except if the window itself is destroyed.

robertknight replied:
`FrameSyncService` currently provide any way to destroy/dispose it. This is true of all services in the Hypothesis client as they are instantiated only once when the application starts. In unit tests we pass in a fake window which will be destroyed at the end of each test. The same applies to other services passed into the constructor.

esanzgar reviewed (commented):
The approach suggested here is sensible transition. @robertknight  has well identified all the limitations

A left a few comments.

robertknight commented on src/annotator/index.js:
```diff
@@ -68,28 +68,32 @@ function init() {
-        ?.sidebarWindow;
+  const sidebarWindow =
+    window_.__hypothesis.sidebarWindow ||
+    /** @type {HypothesisWindow} */ (window.parent).__hypothesis?.sidebarWindow;
```
Currently only guest frames which direct children of the host frame and same-origin are supported. In this case `window.__hypothesis.sidebarWindow` is assigned in the host frame and is accessed via `window.parent.__hypothesis?.sidebarWindow` in the guest.

esanzgar replied:
Thanks for the clarification.

esanzgar replied:
Because the same code runs on the host frame and on the guest frame/s, it maybe  useful to add a comment or use a if/else indicating in which condition happens what assignment.

For example:
```
let sidebarWindow;
if (!annotatorConfig.subFrameIdentifier) {
// host
sidebarWindow = window.__hypothesis.sidebarWindow
} else {
// guest
sidebarWindow = window.parent.__hypothesis?.sidebarWindow
}
```

It is just a suggestion.

robertknight commented on src/annotator/index.js:
```diff
@@ -68,28 +68,32 @@ function init() {
+    );
   } else {
     // eslint-disable-next-line no-console
     console.warn(
```
This condition shouldn't be reached in normal usage, but could be reached if the Hypothesis client was manually loaded into a cross-origin guest frame with `{ subFrameIdentifier: '...' }` in the configuration. It could also be reached if somehow Hypothesis was unloaded in the host frame before it had loaded in the guest frame.

esanzgar replied:
I think on the first scenario, a cross-origin guest, this condition won't be reached because it will throw an error earlier trying to to access (window.parent).__hypothesis.

I didn't thought about the second scenario, Hypothesis unloaded in the host frame before it had loaded in the guests frame.

robertknight replied:
> I think on the first scenario, a cross-origin guest, this condition won't be reached because it will throw an error earlier trying to to access (window.parent).__hypothesis.

You're right. I thought that `window.parent` would be a different kind of `Window` object in a cross-origin context that would just appear to have fewer properties and not expose any globals, so `window.parent.__hypothesis` would return `undefined`.

However this is not what happens. It does indeed generate an exception:

<img width="1029" alt="window-parent-access" src="https://user-images.githubusercontent.com/2458/127309776-a9d206d9-c7cc-43a0-866b-7dcd8b2d04bd.png">

robertknight replied:
I pushed a commit to improve this. See https://github.com/hypothesis/client/pull/3611#issuecomment-888216575.

robertknight commented on src/annotator/cross-frame.js:
```diff
@@ -71,26 +69,26 @@ export class CrossFrame {
-        );
-      });
+    this.connectToSidebar = (frame, origin) => {
+      const channel = new MessageChannel();
```
It does save a few characters. I'm not sure whether its any more or less readable. I left as-is for the moment.

robertknight commented:
Thanks for the feedback @esanzgar. I removed the `port` property from the `hypothesisGuestReady` message in favor of using `MessageEvent.ports`, per your suggestion. I also fixed an issue with a test that was causing code coverage misses.

This is now ready for review.

robertknight commented:
I pushed a commit to improve the handling of the situation where a guest-only frame has a cross-origin parent. Before it triggered an error. Now it outputs a useful console warning:

<img width="824" alt="cross-origin-parent" src="https://user-images.githubusercontent.com/2458/127311335-604743c3-ae7d-4909-bb11-151b88cbb201.png">

esanzgar commented on src/annotator/test/sidebar-test.js:
```diff
@@ -155,6 +155,27 @@ describe('Sidebar', () => {
 
+  describe('#ready', () => {
+    it('returns a promise that resolves when `hypothesisSidebarReady` message is received', async () => {
+      const sidebar = createSidebar();
+
+      // Check `sidebar.ready` is not resolved before `hypothesisSidebarReady`
+      // message is received.
+      assert.equal(
+        await Promise.race([sidebar.ready, Promise.resolve('not-ready')]),
+        'not-ready'
+      );
+
+      window.dispatchEvent(
+        new MessageEvent('message', {
+          data: { type: 'hypothesisSidebarReady' },
+        })
+      );
+
+      return sidebar.ready;
```
I found this test a bit contrive. What about this alternative?
```suggestion
      const sidebar = createSidebar();

      window.dispatchEvent(
        new MessageEvent('message', {
          data: { type: 'hypothesisSidebarReady' },
        })
      );
      await sidebar.ready;

      assert(true);
```

Or maybe:

```suggestion
      const sidebar = createSidebar();
      let counter = 0;
      const promise = sidebar.ready.then(() => ++counter);

      window.dispatchEvent(
        new MessageEvent('message', {
          data: { type: 'hypothesisSidebarReady' },
        })
      );
      await promise;

      assert.equal(counter, 1);
```

robertknight replied:
The `Promise.race` part is an attempt to check that the `sidebar.ready` Promise is not already resolved. JavaScript intentionally doesn't provide APIs to inspect Promise state, so the best we can do is race it against a Promise that is immediately resolved.

```
 const promise = sidebar.ready.then(() => ++counter);
 ```
 
Promises can only be resolved once, so unless `sidebar.ready` were re-assigned to a different Promise, `counter` can never have a value that is more than one here.

esanzgar reviewed (approved):
It looks good.

I left a couple of very minor suggestions.

robertknight commented on src/annotator/test/cross-frame-test.js:
```diff
@@ -81,34 +57,37 @@ describe('CrossFrame', () => {
-        origin: 'ORIGIN',
-        token: 'TOKEN',
-      });
+      cf.connectToSidebar(sidebarFrame);
```
Will do.

robertknight commented on src/annotator/index.js:
```diff
@@ -68,28 +68,39 @@ function init() {
+  let sidebarWindow = window_.__hypothesis.sidebarWindow;
+  try {
+    sidebarWindow = /** @type {HypothesisWindow} */ (window.parent).__hypothesis
+      ?.sidebarWindow;
```
I've added a comment here. More importantly looking at this again made be realize there was a bug where the assignment in the `try` block did not check that `sidebarWindow` was not already assigned.
